internal/pkg/redis/streams/client.go

```diff
@@ -0,0 +1,200 @@
+/********************************************************************************
+ *  Copyright 2020 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package streams
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/edgexfoundry/go-mod-messaging/internal/pkg"
+	"github.com/edgexfoundry/go-mod-messaging/pkg/types"
+)
+
+// Client MessageClient implementation which provides functionality for sending and receiving messages using
+// RedisStreams.
+type Client struct {
+	// Client used for functionality related to reading messages
+	subscribeClient RedisClient
+
+	// Client used for functionality related to sending messages
+	publishClient RedisClient
+}
+
+// NewClient creates a new Client based on the provided configuration.
+func NewClient(messageBusConfig types.MessageBusConfig) (Client, error) {
+	return NewClientWithCreator(messageBusConfig, NewGoRedisClientWrapper, tls.X509KeyPair, tls.LoadX509KeyPair)
+}
+
+// NewClientWithCreator creates a new Client based on the provided configuration while allowing more control on the
+// creation of the underlying entities such as certs, keys, and Redis clients
+func NewClientWithCreator(
```

Comment:
Is this intended to be a singleton?

Reply:
No, my original intent was to have this function along with the `NewClient` function to create new clients for each call, so that we can get over some of the "client" limitations, like only be able to subscribe and/or publish to one host per client.